test/python/pulse/test_builder.py

```diff
@@ -0,0 +1,950 @@
+                pulse.call_circuit(twice_cx_qc)
+        self.assertNotEqual(len(schedule.instructions), 0)
+
+        with pulse.build(self.backend) as schedule:
+            with pulse.transpiler_settings(optimization_level=3):
+                pulse.call_circuit(twice_cx_qc)
+        self.assertEqual(len(schedule.instructions), 0)
+
+    def test_scheduler_settings(self):
+        """Test the circuit scheduler settings context."""
+        inst_map = pulse.InstructionScheduleMap()
+        d0 = pulse.DriveChannel(0)
+        test_x_sched = pulse.Schedule()
+        test_x_sched += instructions.Delay(10, d0)
+        inst_map.add('x', (0,), test_x_sched)
+
+        x_qc = circuit.QuantumCircuit(2)
+        x_qc.x(0)
+
+        with pulse.build(backend=self.backend) as schedule:
+            with pulse.transpiler_settings(basis_gates=['x']):
+                with pulse.circuit_scheduler_settings(inst_map=inst_map):
+                    pulse.call_circuit(x_qc)
+
+        self.assertEqual(schedule, test_x_sched)
+
+    def test_phase_offset(self):
+        """Test the phase offset context."""
+        d0 = pulse.DriveChannel(0)
+
+        with pulse.build() as schedule:
+            with pulse.phase_offset(3.14, d0):
+                pulse.delay(10, d0)
+
+        reference = pulse.Schedule()
+        reference += instructions.ShiftPhase(3.14, d0)
+        reference += instructions.Delay(10, d0)
+        reference += instructions.ShiftPhase(-3.14, d0)
+
+        self.assertEqual(schedule, reference)
+
+    @unittest.expectedFailure
+    def test_frequency_offset(self):
+        """Test the frequency offset context."""
+        d0 = pulse.DriveChannel(0)
+
+        with pulse.build() as schedule:
+            with pulse.frequency_offset(1e9, d0):
+                pulse.delay(10, d0)
+
+        reference = pulse.Schedule()
+        reference += instructions.ShiftFrequency(1e9, d0)  # pylint: disable=no-member
+        reference += instructions.Delay(10, d0)
+        reference += instructions.ShiftFrequency(-1e9, d0)  # pylint: disable=no-member
+
+        self.assertEqual(schedule, reference)
+
+    @unittest.expectedFailure
+    def test_phase_compensated_frequency_offset(self):
+        """Test that the phase offset context properly compensates for phase accumulation."""
+        d0 = pulse.DriveChannel(0)
+
+        with pulse.build(self.backend) as schedule:
+            with pulse.frequency_offset(1e9, d0, compensate_phase=True):
+                pulse.delay(10, d0)
+
+        reference = pulse.Schedule()
+        reference += instructions.ShiftFrequency(1e9, d0)  # pylint: disable=no-member
+        reference += instructions.Delay(10, d0)
+        reference += instructions.ShiftPhase(-1e9*10/self.configuration.dt, d0)
+        reference += instructions.ShiftFrequency(-1e9, d0)  # pylint: disable=no-member
+
+        self.assertEqual(schedule, reference)
+
+
+class TestTypes(TestBuilder):
```

Comment:
I'm still confused by how you are using "types" in this PR

Reply:
The type of `DriveChannel(0)` is a `DriveChannel`, and also `PulseChannel`, and `Channel`, but I wouldn't say `DriveChannel(0)` is itself a type

Reply:
You're right I've fixed in 92fd872d3